// rollup-plugin-auto-reload/dist/index.js
'use strict';

var fs = require('fs');
var http = require('http');
var MagicString = require('magic-string');
var path = require('path');
var ws = require('ws');

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol */


function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

const autoReload = (init = {}) => {
    const config = Object.assign({ port: 0, host: "localhost" }, init);
    let watch = false;
    let httpServer = undefined;
    let webSocketServer = undefined;
    return {
        name: "Rollup-Plugin-Auto-Reload",
        options(options) {
            return __awaiter(this, void 0, void 0, function* () {
                watch = !!options.watch;
            });
        },
        buildStart() {
            return __awaiter(this, void 0, void 0, function* () {
                if (watch && !httpServer) {
                    httpServer = http.createServer().listen(config.port);
                    webSocketServer = new ws.WebSocketServer({ server: httpServer });
                }
            });
        },
        buildEnd() {
            return __awaiter(this, void 0, void 0, function* () {
                if (watch) {
                    setTimeout(() => webSocketServer === null || webSocketServer === void 0 ? void 0 : webSocketServer.clients.forEach(socket => socket.send("RELOAD")));
                }
            });
        },
        closeWatcher() {
            return __awaiter(this, void 0, void 0, function* () {
                webSocketServer === null || webSocketServer === void 0 ? void 0 : webSocketServer.close();
                httpServer === null || httpServer === void 0 ? void 0 : httpServer.close();
            });
        },
        transform(code, module) {
            return __awaiter(this, void 0, void 0, function* () {
                if (watch && httpServer && this.getModuleInfo(module).isEntry) {
                    const address = httpServer.address();
                    const path$1 = path.resolve(__dirname, "client.js");
                    const supplement = fs.readFileSync(path$1).toString()
                        .replaceAll("HOST", config.host)
                        .replaceAll("PORT", address.port.toString());
                    return {
                        code: supplement + code,
                        map: new MagicString(supplement + code).generateMap({ source: "client.js", includeContent: true }),
                    };
                }
                else {
                    return {
                        code,
                        map: this.getCombinedSourcemap(),
                    };
                }
            });
        },
    };
};

exports.autoReload = autoReload;
